### PR TITLE
More data

### PR DESCRIPTION
**Pull request by ghost**
more data from html, Ajax & api. api did have some attributes that were wrong or should have been nullable but weren't. I tested it on [output.txt](https://github.com/user-attachments/files/18229903/output.txt)

The ids were scraped from the different DLsite home pages(upcoming, r18, all ages, game,...) & extracted with regex.

### CODE REVIEW

File: src/product/ajax.rs [hunk 1]
```diff
@@ -4,9 +4,199 @@ use serde::{Deserialize, Deserializer};
...
+    pub workno: String,
+    pub id: i32,
+    pub start_date: String,
+    pub original_workno: Option<String>,
+    pub end_date: String,
+}
+
 /// Data of a product from the AJAX API.
 #[derive(Debug, Clone, Deserialize)]
+#[serde(deny_unknown_fields)]
```
Comment (ozonezone) [inline src/product/ajax.rs]: I think request should not be failed even if new field is added. Maybe it is good to add deny_unknown_fields attribute only when testing.

File: src/product_api/interface.rs [hunk 2]
```diff
@@ -281,9 +407,16 @@ pub struct Content {
     #[serde(rename = "upper(work_files.type)")]
     pub upper_work_files_type: String,
     pub extension: String,
+    name_url: Option<String>,
+    title: Option<String>,
+    filesize: Option<String>,
+    name: Option<String>,
+    filepath: Option<String>,
+    size: Option<String>,
```
Comment (ozonezone) [inline src/product_api/interface.rs]: Is there any reason to these fields private?

File: src/product_api/interface.rs [hunk 3]
```diff
@@ -306,17 +439,38 @@ pub struct File {
...
+    file_size_unit: Option<String>,
+    width: Option<String>,
+    path_short: Option<String>,
+    #[serde(rename = "upper(work_files.type)")]
+    upper_work_files_type: Option<String>,
+    #[serde(rename = "type")]
+    type_: Option<String>,
+    resize_url: Option<String>,
+    workno: Option<String>,
+    extension: Option<String>,
```
Comment (ozonezone) [inline src/product_api/interface.rs]: Same as above

File: src/product_api/interface.rs [hunk 4]
```diff
@@ -6,8 +6,106 @@ use serde_with::{formats::PreferOne, serde_as, DefaultOnError, OneOrMany};
...
+pub type StrOrBool = TwoPossibility<String, bool>;
+pub type ArrOrSingle<T> = TwoPossibility<Vec<T>, T>;
+pub type HashMapOrArr<T> = TwoPossibility<HashMap<String, T>, Vec<T>>;
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(untagged)]
+pub enum TwoPossibility<T, V> {
+    One(T),
+    Two(V),
+}
```
Comment (ozonezone) [inline src/product_api/interface.rs:14]: I think `either` crate can be used instead of defining own type but this maybe enough for now.

**Review by ozonezone — COMMENTED** [review]
Hi, thanks for contribution! I've run random test several times and it seems good.
I have some review comments. Could you check it?

File: src/product/ajax.rs (hunk 1 above)
Comment (ghost) [reply src/product/ajax.rs]: yeah makes sense. what do you think about https://docs.rs/serde_ignored/latest/serde_ignored/

File: src/product_api/interface.rs (hunk 3 above)
Comment (ghost) [reply src/product_api/interface.rs]: forgot to set the pub

**Comment by ghost** [comment]
the version bump broke the  self
            .get(&format!("/api/=/product.json?workno={}", id))
            .await?; the return is always "[]". not quite sure why. I have. a look into it

File: src/product/ajax.rs (hunk 1 above)
Comment (ozonezone) [reply src/product/ajax.rs]: That looks good, but for now it will be enough to just enable the `strict` feature in ci.

**Comment by ghost** [comment]
it fixed itself, not I am trying to figure out what's the issue with either. maybe the api was down the moment I run the test

**Comment by ozonezone** [comment]
I guess `either::serde_untagged` is needed? But just reverting either is okay.

**Comment by ghost** [comment]
I reverted it back bc either::serde_untagged would need to used every time a variable uses Either. I bumped the versions including the version of this crate & added the url to the repo

**Comment by ghost** [comment]
also added serde_ignored as a feature. it will not warn, but log it as an error to increase the probability that it will be reported

**Comment by ozonezone** [comment]
LGTM, thanks very much! Note that feature name is changed.

**Pull request merged**